src/utils/ahash.zig

```diff
@@ -0,0 +1,228 @@
+const std = @import("std");
+const sig = @import("../sig.zig");
+
+const AtomicU64 = std.atomic.Value(u64);
+const Random = std.rand.Random;
+
+/// AHasher is a minimal impelementation of the AHash, a fast, DOS-resistant hash algorithm (https://github.com/tkaitchuck/aHash).
+/// Currently, only the functionality required to replicate the agave Deduper is implemented.
+pub const AHasher = struct {
+    buffer: u64,
+    pad: u64,
+    extra_keys: [2]u64,
+
+    // This constant comes from Kunth's prng (Empirically it works better than those from splitmix32).
+    const MULTIPLE: u64 = 6364136223846793005;
+    const ROTATE: u32 = 23;
+
+    pub fn fromSeed(seed: AHashSeed) AHasher {
+        return .{
+            .buffer = seed.k0,
+            .pad = seed.k1,
+            .extra_keys = .{ seed.k2, seed.k3 },
+        };
+    }
+
+    pub fn hash(self: *AHasher, comptime T: type, data: *const T) void {
+        switch (@typeInfo(T)) {
+            .Int => {
+                self.update(@intCast(data.*));
+            },
+            .Array => |array_info| {
+                self.hashSlice(array_info.child, data);
+            },
+            .Pointer => |pointer| {
+                switch (pointer.size) {
+                    .Slice => self.hashSlice(pointer.child, data.*),
+                    else => {
+                        std.debug.print(
+                            "Unsupported pointer size: type={} pointer.size={}\n",
+                            .{ T, pointer.size },
+                        );
+                        std.debug.panic(
+                            "Unsupported pointer size: type={} pointer.size={}\n",
+                            .{ T, pointer.size },
+                        );
+                    },
+                }
+            },
+            .Struct => |struct_info| {
+                inline for (struct_info.fields) |field| {
+                    self.hash(field.type, &@field(data, field.name));
+                }
+            },
+            .Enum => |_| {
+                self.hash(u32, &@as(u32, @intFromEnum(data.*)));
+            },
+            else => {
+                std.debug.print("Unsupported type: type={} data={any}\n", .{ T, data });
+                std.debug.panic("Unsupported type: type={} data={any}\n", .{ T, data });
+            },
+        }
+    }
+
+    pub fn finish(self: *const AHasher) u64 {
+        const rot: u32 = @intCast(self.buffer & 63);
+        return std.math.rotl(u64, foldedMultiply(self.buffer, self.pad), rot);
+    }
+
+    inline fn hashSlice(self: *AHasher, comptime T: type, data: []const T) void {
+        self.update(data.len);
```

Comment:
What I meant was it seems unnecessary to hash the length.

Reply:
Not really, if you don't hash the length then there is no distinction between hashing [a, b, c] and hashing [a] then hashing [b, c]. You could argue that this distinction doesn't matter for this use case and perhaps I'd agree but to be honest I don't think it is enough of an issue to bother breaking dependent unit tests for right now. Perhaps it can be revisited in the perf task which I have added to our backlog

Reply:
ok that's a good point, i'm fine with how it is